File: src/aoc_2024/day5.clj
(ns aoc-2024.day5
  (:require [clojure.string :as str]))

(def input (slurp "resources/day5.txt"))

(defn parse [pred s]
  (->> (str/split input #"\r\n\r\n")
       (pred)
       (str/split-lines)
       (mapv #(mapv read-string (str/split % s)))))

(def orderings (parse first #"\|"))

(def pages (parse second #","))

(defn middle [xs] (nth xs (quot (count xs) 2)))

(defn page-comp [x y]
  (cond
    (.contains orderings [x y]) -1
    (.contains orderings [y x]) 1
    :else 0))

(defn p-sorted? [p] (= (sort page-comp p) p))

(defn part-1 []
  (->> (filter p-sorted? pages)
       (map middle)
       (reduce +)))

(defn part-2 []
  (->> (remove p-sorted? pages)
       (map (comp middle (partial sort page-comp)))
       (reduce +)))

File: python/2024/day5.py
from functools import cmp_to_key


def comp(x, y):
    if [x, y] in orderings: #type: ignore
        return -1 
    elif [y, x] in orderings: #type: ignore
        return 1 
    else:
        return 0

with open("day5.txt", "r") as infile:
    data = infile.read()
    orderings, pages = data.split("\n\n")
    orderings = [[int(x) for x in s.split()] for s in orderings.splitlines()]
    pages = [[int(x) for x in s.split(",")] for s in pages.splitlines()]
    total = 0
    total_p2 = 0

    for entry in pages:
        entry_sorted = sorted(entry, key=cmp_to_key(comp))
        if entry_sorted == entry:
            total += entry[len(entry) // 2]
        else:
            total_p2 += entry_sorted[len(entry) // 2]


    print(total)
    print(total_p2)


